tools/autograd/gen_variable_type.py

```diff
@@ -292,10 +296,57 @@
 #endif
 """)
 
+FW_DERIVATIVE_CHECK_TEMPLATE = CodeTemplate("""\
+isFwGradDefined(${req_inp})\
+""")
```

Comment:
FWIW, the modern codegen style (we're not uniformly here yet) is to put single use templates inline, as it makes it easier to tell what code is being generated at any given point in time.

Reply:
To take `FW_DERIVATIVE_CHECK_TEMPLATE` as an example, there is negative utility to having it out of line here: it's not a multi-line substitution and it is only used in one place. Format strings are just unconditionally better.

Reply:
Could you be more precision on how to inline these nicely?

In here for example that leads to
```python
elif isinstance(derivative.var_type, ListType) and derivative.var_type.is_tensor_like():
                fw_grad_setter = f'TORCH_INTERNAL_ASSERT(${res}.size() == ${res}_new_fw_grad.size());'
                fw_grad_setter += f'for (auto i=0; i<${res}.size(); ++i) {{'
                fw_grad_setter += f'  if (${res}_new_fw_grad[i].defined()) {{'
                fw_grad_setter += f'    ${res}[i].set_fw_grad(${res}_new_fw_grad[i], /* level / 0, / is_inplace_op */ ${is_inplace_str});'
                fw_grad_setter += '  }'
                fw_grad_setter += '}'
```

While it is nice to have it inline indeed, it makes the code block harder to read compared to a single block definition like:
```python
FW_DERIVATIVE_SETTER_TENSOR_LIST = CodeTemplate("""\
TORCH_INTERNAL_ASSERT(${out_arg}.size() == ${out_arg}_new_fw_grad.size());
for (auto i=0; i<${out_arg}.size(); ++i) {
  if (${out_arg}_new_fw_grad[i].defined()) {
    ${out_arg}[i].set_fw_grad(${out_arg}_new_fw_grad[i], /* level / 0, / is_inplace_op */ ${is_inplace});
  }
}
""")
```

Am I missing a better way to inline these? Or you were mentioning this only for the one-liner ones?

Reply:
If we had an UNINDENT function (that looks at the minimum indentation level and removes it from each line), then we can inline these nicely:
```python
elif isinstance(derivative.var_type, ListType) and derivative.var_type.is_tensor_like():
    fw_grad_setter = UNINDENT(f"""\
        TORCH_INTERNAL_ASSERT(${res}.size() == ${res}_new_fw_grad.size());
        for (auto i=0; i<${res}.size(); ++i) {
            if (${res}_new_fw_grad[i].defined()) {
                ${res}[i].set_fw_grad(${res}_new_fw_grad[i], /* level / 0, / is_inplace_op */ ${is_inplace_str});
            }
        }""")
        
```